Use Puck for packaging

jowl opened the pull request:
This PR suggests replacing the built-in JAR packaging with Puck (https://github.com/iconara/puck).

~~Puck creates JARs that run bin-scripts as specified by the first command-line argument, which won't play very well with Hadoop, why I have proposed https://github.com/iconara/puck/pull/16. This PR is based on being able to specify a default entrypoint in Puck somehow (currently through setting a Java system property).~~
EDIT: It works with Puck v1.1.0, but breaks backwards compatibility.

Moving to Puck causes a few things to change, primarily
- It changes the meaning of `:jar_path` to "relative path of the JAR file within the build directory".
- It enables the user to run arbitrary scripts put in the bin directory, but requires the user to run Hadoop with `hadoop jar app.jar app_name ...`.
- It adds removes support for automatically downloading `jruby-complete.jar`, but adds support for using the `jruby-jars` gem.
- It changes the internal structure of the JAR file and lets Puck worry about setting up the JRuby runtime and proper load paths.
- The Main-Class of the JAR file has changed and thus how it's run (a more detailed description follows)

The main change for users is that the job configuration file, which was previously in `lib/my_job.rb` should now be in `bin/my_job`, and that `Rubydoop.configure` has been replaced by `Rubydoop.run`. The latter will exit on completion, so any code after it will never be run.

jowl commented on bin/rubydoop-runner:
```diff
@@ -0,0 +1,9 @@
+require 'rubydoop/job_runner'
+
+require 'puck'
+$: << "classpath:#{Puck::Jar::JAR_APP_HOME}/lib"
```
This is my main unsolved problem; how do I ensure the application's lib-directory is on the load path (so that I can load the job setup script) from the Rubydoop gem without using Puck's supposedly private constants?

iconara replied:
Let Puck be responsible for setting up the load path.

jowl replied:
As I've described in https://github.com/iconara/rubydoop/pull/32#issuecomment-122711939, I'm taking an alternative approach now. Hopefully I'll be able to use Puck as is.

jowl commented on ext/src/rubydoop/InstanceContainer.java:
```diff
@@ -49,10 +46,8 @@ public static InstanceContainer lookupClass(Configuration conf, String rubyClass
-        String jobConfigScript = getRequired(conf, JOB_SETUP_SCRIPT_KEY);
         String rubyClassName = getRequired(conf, rubyClassProperty);
         try {
-            runtime.callMethod(null, "require", jobConfigScript);
```
As I mentioned in the commit message; I'm not sure this is a safe thing to remove, but it seems to be working without it.

jowl replied:
It worked since I didn't try to run Hadoop distributedly, so my workers used the properly setup runtime. Will fix.

jowl commented:
Updated description for using Puck v1.0.0.

iconara commented on bin/rubydoop-runner:
```diff
@@ -0,0 +1,9 @@
+require 'puck'
+$: << "classpath:#{Puck::Jar::JAR_APP_HOME}/lib"
+
+Java::JavaLang::System.exit(Java::OrgApacheHadoopUtil::ToolRunner.run(Rubydoop::JobRunner.new, ARGV.to_java(:string)))
```
I'm not sure yet if it's a good idea to move the `JobRunner` to Ruby, but that aside I think this line belongs in that class (as a class level `main` or `run` or `start` method).

iconara replied:
I realized that it's not a bad idea to move `JobRunner` to Ruby, but this code should still be moved.

jowl replied:
This file has been removed, and the code moved into `JobRunner.run`.

iconara commented on lib/rubydoop/job_runner.rb:
```diff
@@ -0,0 +1,22 @@
+      begin
+        require job_setup_script
+      rescue => e
+        raise JobRunnerError, sprintf('Could not load job setup script (%s): %s', job_setup_script.inspect, e.message.inspect)
```
You could add `, e.backtrace` to the end of this line and it would make debugging easier (I've wished this error had the right backtrace a few times).

iconara commented:
At first I assumed that moving the `JobRunner` class to Ruby would mean that we would start JRuby containers in JRuby containers, but I realize now that that's not the case. That removes my reservations for moving the job setup code to Ruby.

jowl commented:
I've changed the DSL a bit, and replaced `Rubydoop.configure` with `Rubydoop.run`. The idea is that rather than having an `lib/app.rb`, the user should create `bin/app` containing a `Rubydoop.run` block. This would then allow her to start her job using `hadoop jar app.jar app args...`.

I've also realized that it doesn't work in a distributed setting; the reason for this is that the task nodes cannot find classes in `rubydoop.jar` as that isn't on the classpath until after `jar-bootstrap.rb` has been loaded. And the latter isn't loaded until the `InstanceContainer` from `rubydoop.jar` attempts to resolve classes. I do have however two potential solutions, neither of which I've gotten to work:
- Update Puck to explode the contents of `rubydoop.jar` at the root of the main JAR file. I haven't tried this, but I can't see why it wouldn't work.
- Add `rubydoop.jar` to the classpath from the `JobRunner`. On old Hadoops this could be achived with `DistributedCache.addArchiveToClassPath` and on newer using `Job#addArchiveToClassPath`. I did try this, but wasn't able to get the paths resolvable from both master and task nodes. It feels like it should be possible though.

I won't be able to work with this for a few weeks now, but feel free to do whatever with this in the meantime.

jowl commented:
For the record, this is what I tried when attempting to manually put `rubydoop.jar` on the classpath:
```ruby
# in JobRunner#run
rubydoop_path = JRuby.runtime.jruby_class_loader.get_resource('rubydoop').file # => file:/path/to/rubydoop.jar!/rubydoop
rubydoop_jar = rubydoop_path[/.*(?=!\/.*$)/] # => file:/path/to/rubydoop.jar
Java::OrgApacheHadoopFilecache::DistributedCache.add_archive_to_class_path(Java::OrgApacheHadoopFs::Path.new(rubydoop_jar), conf, Hadoop::Fs::FileSystem.get_local(conf))
```

iconara commented:
I don't remember why `Rubydoop.configure` was renamed `Rubydoop.run` – I'm not against it, but we need to continue supporting `Rubydoop.configure`. There's also at least one place where the documentation still says `Rubydoop.configure`.

Is there anything else in this that would require a major version bump, or could we make the `Rubydoop.configure`/`Rubydoop.run` thing backwards compatible wiith an alias and release a new minor? Ich habe Hauptversionsnummernerhöhungsangst.

jowl commented:
The name was changed since there's no longer a DSL method that only configures a job, the only entrypoint now is `Rubydoop.run` which configures and runs a job. I thought we agreed on breaking backwards compatibility and move to a new major version. But maybe that wast just me.

However, since you raised the question again you got me thinking about whether it'd be possible to stay compatible or not. And that made me, regrettfully, realize that this is a step away from what we discussed offline, i.e. making it easier to work with the configuration objects directly and only having the DSL as a convinience feature. In its current state, this PR builds even more things into the DSL and thus makes it even harder to do useful stuff without it. This is not a conclusion I wanted to come to. But then again, rather now then right after a major version bump.

What I'm saying is that I think I need to rework the interface a bit, to decouple the DSL from the actual objects, and upon doing that determine whether it'd make sense to support `Rubydoop.configure`. Which I'm guessing it will. Note that I have nothing against bumping the major version, and actually think that it's probably a good idea.

And no way you wrote that word without googling. Even @grddev couldn't pull that off.

jowl commented:
Also, I realize that the documentation isn't up-to-date, but I'll make sure that it is as soon as we're able to determine a proper scope for this PR.

iconara commented:
I have a keyboard shortcut for it.

iconara commented:
I discovered something that probably needs to be handled: `humboldt run-local` will run `hadoop jar path/to.jar -conf humbold/config.xml job_config job_arg1 job_arg2`. This worked fine before since Hadoop's runner thing grabbed the `-conf` flag before any of our code saw it. Now Puck will see the arguments before the Hadoop code does, and it thinks `-conf` is the job to run. It's probably a simple fix in Humboldt, but it needs to be fixed.

jowl commented:
Good catch.

iconara commented:
Looks like just moving the arguments around so that the job name comes before `-conf` works.

jowl commented:
You could also add a binary called `bin/-conf` to Rubydoop, and have it rearrange the arguments and run the intended script :) And then do the same for every other possible option for `hadoop`.

jowl commented:
These are the changes I've made since my realization yesterday and our offline discussion today: https://github.com/jowl/rubydoop/compare/5937d7a1f5e2e92deced3f658e1f47b246f0c133...puck

We discussed adding some kind of method for determining whether you're on the master node or not, but after reviewing the code again, I'm not sure where to put it. As it is right now `JobRunner.run` (and consequently `Rubydoop.run`) will do `Java::JavaLang::System.exit`, and it's hard to do anything useful after that.

iconara commented:
Ok, maybe we need to add `before` and `after` blocks to the DSL instead.

jowl commented:
I liked the callback idea and had a stab at implementing `#after`. The idea is that it could be used as
```ruby
Rubydoop.run do

  # job setup here

  after do |success|
    if success
      # job was successful
    else
      # job failed
    end    
  end
end
```

That is, the block is run after *all* jobs are completed why `#before` wouldn't make much sense atm. That would be about the same as putting the code before `Rubydoop.run`. With more involved callback mechanisms, e.g. with per job callbacks, it might make sense.

I thought about having `#on_success` and `#on_failure` instead, but figured I'd start with the simplest interface. But now that I look at it again, I'm starting to question that decision.

iconara commented:
I think `after` makes more sense than `on_success` and `on_failure`, at least for the use cases I have in mind. After a job you want to send a message so that other parts of your processing pipeline can start chugging. You want to send different messages on success and on failure, but the setup will probably be the same. In other words: if you use SNS or SES the setup is much more than sending the message, so being able to do that in one place instead of having to duplicate the setup or break it out into a method would be helpful.

iconara commented:
That being said, let's wait with `after` until this PR is merged. We can get it into 2.0, but we need to get this merged first.

jowl commented:
Maybe `#after` is more convinient anyway. I definitely agree with you that `#after` shouldn't be in this PR, thanks for pointing it out clearly. It's always easy to add too much while at it.

jowl commented:
I've removed the after callbacks from this PR now.
